### ISSUE TITLE
Http4sMatchers gives a test failure for haveHeaders when the headers are correct

### ISSUE THREAD

**Issue by McBrainy** [http4s/http4s#2099]
[One of your tests](https://github.com/http4s/http4s/blob/master/dsl/src/test/scala/org/http4s/dsl/ResponseGeneratorSpec.scala#L102), which presumably passes since your build is green (I haven't run it myself), fails if you copy its source code into a test case using Http4sMatchers, with a somewhat bizarre error. Here's the error:

```
[info] HelloWorldSpec
[error] x testing
[error]  the headers 'Headers(Content-Length: 0, Location: http://foo, Accept: audio/*)' is not the same as Headers(Content-Length: 0, Location: http://foo, Accept: audio/*) (HelloWorldSpec.scala:15)
[info] Total for specification HelloWorldSpec
[info] Finished in 1 second, 235 ms
[info] 1 example, 1 failure, 0 error
```

The source:

```scala
package com.example.http4stmp

import cats.effect.IO
import org.http4s._
import org.http4s.dsl.io._
import org.http4s.headers._
import org.http4s.testing.Http4sMatchers
import org.specs2.mutable.Specification

class HelloWorldSpec extends Specification with Http4sMatchers {

  "testing" >> {
    val location = Location(Uri.unsafeFromString("http://foo"))
    val resp = MovedPermanently(location, Accept(MediaRange.`audio/*`))
    resp must returnValue(
      haveHeaders(
        Headers(
          `Content-Length`.zero,
          location,
          Accept(MediaRange.`audio/*`)
        )))
  }

}
```

The build.sbt:

```scala
val Http4sVersion = "0.18.18"
val Specs2Version = "4.2.0"
val LogbackVersion = "1.2.3"

lazy val root = (project in file("."))
  .settings(
    organization := "com.example",
    name := "http4s-tmp",
    version := "0.0.1-SNAPSHOT",
    scalaVersion := "2.12.6",
    libraryDependencies ++= Seq(
      "org.http4s"      %% "http4s-blaze-server" % Http4sVersion,
      "org.http4s"      %% "http4s-circe"        % Http4sVersion,
      "org.http4s"      %% "http4s-dsl"          % Http4sVersion,
      "org.http4s"      %% "http4s-testing"          % Http4sVersion,
      "org.specs2"     %% "specs2-core"          % Specs2Version % "test",
      "ch.qos.logback"  %  "logback-classic"     % LogbackVersion
    ),
    addCompilerPlugin("org.spire-math" %% "kind-projector"     % "0.9.6"),
    addCompilerPlugin("com.olegpy"     %% "better-monadic-for" % "0.2.4")
  )
```

[Related SO question.](https://stackoverflow.com/questions/52416411/http4s-what-is-difference-between-actual-and-expected-value)

**Comment by rossabaker** [comment]
Interesting.  This works, which is not far from what that matcher is doing.

```scala
    val location = Location(Uri.unsafeFromString("http://foo"))
    val resp = MovedPermanently(location, Accept(MediaRange.`audio/*`))
    resp.unsafeRunSync().headers must_== Headers(
      `Content-Length`.zero,
      location,
      Accept(MediaRange.`audio/*`)
    )
```

This looks like a bug.

**Comment by rossabaker** [comment]
The test was added on master, and the matcher was fixed.  But it buggily used reference equality in the 0.18 series.  #2101 will fix it.

**Comment by rossabaker** [comment]
Fixed by #2101.
